File: window.go
package main

import (
	"context"
	"fmt"
	"log"
	"sync"
	"syscall"
	"unsafe"

	"github.com/lxn/win"
)

const (
	borderWidth = 2 // Border width in pixels
)

var (
	className  = mustGetUTF16PtrFromString("RedBorderWindow")
	windowName = mustGetUTF16PtrFromString("LiveFrame - YouTube Streaming Border")

	idcArrow = mustGetUTF16PtrFromString("IDC_ARROW")
)

func mustGetUTF16PtrFromString(str string) *uint16 {
	ptr, err := syscall.UTF16PtrFromString(str)
	if err != nil {
		panic(fmt.Sprintf("failed to convert string %q to UTF16 pointer", str))
	}

	return ptr
}

// WindowManager manages the window visibility
type WindowManager struct {
	hwnd win.HWND
	mu   sync.Mutex
}

// NewWindowManager creates a new window manager
func NewWindowManager(hwnd win.HWND) *WindowManager {
	return &WindowManager{
		hwnd: hwnd,
	}
}

// SetVisible sets the window visibility
func (wm *WindowManager) SetVisible(visible bool) {
	if wm == nil {
		log.Println("Warning: windowManager is nil, cannot update visibility")
		return
	}

	wm.mu.Lock()
	defer wm.mu.Unlock()

	// Check if window handle is valid
	if wm.hwnd == 0 {
		log.Println("Warning: Invalid window handle in SetVisible")
		return
	}

	if visible {
		win.ShowWindow(wm.hwnd, win.SW_SHOW)
		win.UpdateWindow(wm.hwnd)
		log.Println("Window is now visible - YouTube stream detected")
	} else {
		win.ShowWindow(wm.hwnd, win.SW_HIDE)
		log.Println("Window is now hidden - No active YouTube stream")
	}
}

// CreateBorderWindow creates the border window that will be shown during streaming
func CreateBorderWindow(ctx context.Context) (win.HWND, *WindowManager, error) {
	// Register window class

	hInstance := win.GetModuleHandle(nil)

	var icex win.INITCOMMONCONTROLSEX
	icex.DwSize = uint32(unsafe.Sizeof(icex))
	icex.DwICC = win.ICC_STANDARD_CLASSES
	win.InitCommonControlsEx(&icex)

	wcex := win.WNDCLASSEX{
		CbSize:        uint32(unsafe.Sizeof(win.WNDCLASSEX{})),
		Style:         win.CS_HREDRAW | win.CS_VREDRAW,
		LpfnWndProc:   syscall.NewCallback(wndProc),
		HInstance:     hInstance,
		HCursor:       win.LoadCursor(0, idcArrow),
		HbrBackground: win.HBRUSH(win.GetStockObject(win.BLACK_BRUSH)),
		LpszClassName: className,
	}

	if atom := win.RegisterClassEx(&wcex); atom == 0 {
		return 0, nil, fmt.Errorf("RegisterClassEx failed")
	}

	// Get screen dimensions
	screenWidth := int32(win.GetSystemMetrics(win.SM_CXSCREEN))
	screenHeight := int32(win.GetSystemMetrics(win.SM_CYSCREEN))

	hwnd := win.CreateWindowEx(
		WS_EX_LAYERED|WS_EX_TOPMOST|WS_EX_NOACTIVATE,
		className,
		windowName,
		win.WS_POPUP,
		0, 0, screenWidth, screenHeight,
		0, 0, hInstance, nil,
	)

	if hwnd == 0 {
		return 0, nil, fmt.Errorf("CreateWindowEx failed")
	}

	// Make window transparent except for the border
	win.SetWindowLong(hwnd, win.GWL_EXSTYLE, win.GetWindowLong(hwnd, win.GWL_EXSTYLE)|WS_EX_LAYERED)

	// Make window transparent
	if !SetLayeredWindowAttributes(hwnd, 0, 0, LWA_COLORKEY) {
		return 0, nil, fmt.Errorf("SetLayeredWindowAttributes failed")
	}

	// Create window manager (initially hidden)
	windowManager := NewWindowManager(hwnd)
	windowManager.SetVisible(false)

	// Clean up when context is done
	go func() {
		<-ctx.Done()

		log.Printf("destroying window due to context cancel")
		win.DestroyWindow(hwnd)
	}()

	return hwnd, windowManager, nil
}

func drawRedBorder(hwnd win.HWND) {
	var rc win.RECT
	win.GetClientRect(hwnd, &rc)

	hdc := win.GetDC(hwnd)
	defer win.ReleaseDC(hwnd, hdc)

	// Create a red brush
	redBrush := CreateSolidBrush(win.RGB(255, 0, 0))
	defer win.DeleteObject(win.HGDIOBJ(redBrush))

	// Select the brush into the DC
	oldBrush := win.SelectObject(hdc, win.HGDIOBJ(redBrush))
	defer win.SelectObject(hdc, oldBrush)

	// Draw the top border
	PatBlt(hdc, 0, 0, int(rc.Right), borderWidth, PATCOPY)

	// Draw the bottom border
	PatBlt(hdc, 0, int(rc.Bottom-borderWidth), int(rc.Right), borderWidth, PATCOPY)

	// Draw the left border
	PatBlt(hdc, 0, 0, borderWidth, int(rc.Bottom), PATCOPY)

	// Draw the right border
	PatBlt(hdc, int(rc.Right-borderWidth), 0, borderWidth, int(rc.Bottom), PATCOPY)
}

func wndProc(hwnd win.HWND, msg uint32, wParam, lParam uintptr) uintptr {
	switch msg {
	case win.WM_DESTROY:
		win.PostQuitMessage(0)
		return 0

	case win.WM_KEYDOWN:
		// Close on ESC key
		if wParam == win.VK_ESCAPE {
			win.DestroyWindow(hwnd)
		}
		return 0

	case win.WM_PAINT:
		var ps win.PAINTSTRUCT
		win.BeginPaint(hwnd, &ps)
		drawRedBorder(hwnd)
		win.EndPaint(hwnd, &ps)
		return 0
	}

	return win.DefWindowProc(hwnd, msg, wParam, lParam)
}

File: src/window.rs
use std::sync::atomic::{AtomicU8, Ordering};
use std::sync::mpsc;
use log::{debug, error, info};
use windows::{
    core::*,
    Win32::Foundation::*,
    Win32::Graphics::Gdi::{
        BeginPaint, CreateSolidBrush, DeleteObject, EndPaint, FillRect, InvalidateRect, PAINTSTRUCT,
    },
    Win32::System::LibraryLoader::GetModuleHandleW,
    Win32::UI::WindowsAndMessaging::*,
};

// Frame color state, read by wndproc in the window thread and written by the audio task.
// 0 = unknown (defaults to red), 1 = red (silent), 2 = green (audible).
pub const COLOR_UNKNOWN: u8 = 0;
pub const COLOR_RED: u8 = 1;
pub const COLOR_GREEN: u8 = 2;

static COLOR_STATE: AtomicU8 = AtomicU8::new(COLOR_UNKNOWN);

// Updates the color state. If the category changed, invalidates the window so wndproc repaints.
pub fn set_color_state(hwnd: HWND, new_state: u8) {
    let prev = COLOR_STATE.swap(new_state, Ordering::Relaxed);
    if prev != new_state && hwnd.0 != 0 {
        unsafe {
            InvalidateRect(hwnd, None, TRUE);
        }
    }
}

// Function to create window and run message loop in a separate thread
pub unsafe fn create_window_and_run_message_loop(tx: mpsc::Sender<HWND>) -> Result<()> {
    // Register the window class
    debug!("Registering window class...");
    let instance = GetModuleHandleW(None)?;
    let window_class = w!("RedFrameWindowClass");

    let wc = WNDCLASSEXW {
        cbSize: std::mem::size_of::<WNDCLASSEXW>() as u32,
        style: CS_HREDRAW | CS_VREDRAW,
        lpfnWndProc: Some(wndproc),
        hInstance: instance,
        hCursor: LoadCursorW(None, IDC_ARROW)?,
        lpszClassName: window_class,
        ..Default::default()
    };

    RegisterClassExW(&wc);

    // Get the dimensions of the main display
    let screen_width = GetSystemMetrics(SM_CXSCREEN);
    let screen_height = GetSystemMetrics(SM_CYSCREEN);
    debug!("Screen dimensions: {}x{}", screen_width, screen_height);

    // Create the window with the specified styles
    info!("Creating window with red frame...");
    let hwnd = CreateWindowExW(
        WS_EX_TOOLWINDOW | WS_EX_LAYERED | WS_EX_TOPMOST | WS_EX_NOACTIVATE,
        window_class,
        w!("Red Frame"),
        WS_POPUP,
        0,             // X position (at the left edge of the screen)
        0,             // Y position (at the top edge of the screen)
        screen_width,  // Width (screen width)
        screen_height, // Height (screen height)
        None,
        None,
        instance,
        None,
    );

    if hwnd.0 == 0 {
        error!("Failed to create window");
        return Err(Error::from_win32());
    }

    // Send the window handle to the main thread
    if let Err(e) = tx.send(hwnd) {
        error!("Failed to send window handle: {}", e);
        return Err(Error::from_win32());
    }

    // Set the window to be transparent except for the red frame
    debug!("Setting window transparency...");
    let color_key = COLORREF(0); // Black is transparent
                                 // Use the full path for SetLayeredWindowAttributes
    SetLayeredWindowAttributes(hwnd, color_key, 255, LWA_COLORKEY);

    // Message loop
    info!("Starting window message loop...");
    let mut message = MSG::default();
    while GetMessageW(&mut message, None, 0, 0).into() {
        TranslateMessage(&message);
        DispatchMessageW(&message);
    }

    info!("Window message loop ended");
    Ok(())
}

extern "system" fn wndproc(hwnd: HWND, message: u32, wparam: WPARAM, lparam: LPARAM) -> LRESULT {
    unsafe {
        match message {
            WM_PAINT => {
                let mut ps = PAINTSTRUCT::default();
                let hdc = BeginPaint(hwnd, &mut ps);

                // Get client area dimensions
                let mut rect = RECT::default();
                GetClientRect(hwnd, &mut rect);

                // COLORREF is 0x00BBGGRR. Green when audio is audible, red otherwise.
                let color = match COLOR_STATE.load(Ordering::Relaxed) {
                    COLOR_GREEN => COLORREF(0x00FF00),
                    _ => COLORREF(0x0000FF),
                };
                let brush = CreateSolidBrush(color);

                let frame_thickness = 3;

                let top_rect = RECT {
                    left: 0,
                    top: 0,
                    right: rect.right,
                    bottom: frame_thickness,
                };
                FillRect(hdc, &top_rect, brush);

                let bottom_rect = RECT {
                    left: 0,
                    top: rect.bottom - frame_thickness,
                    right: rect.right,
                    bottom: rect.bottom,
                };
                FillRect(hdc, &bottom_rect, brush);

                let left_rect = RECT {
                    left: 0,
                    top: 0,
                    right: frame_thickness,
                    bottom: rect.bottom,
                };
                FillRect(hdc, &left_rect, brush);

                let right_rect = RECT {
                    left: rect.right - frame_thickness,
                    top: 0,
                    right: rect.right,
                    bottom: rect.bottom,
                };
                FillRect(hdc, &right_rect, brush);

                DeleteObject(brush);

                EndPaint(hwnd, &ps);
                LRESULT(0)
            }
            WM_DESTROY => {
                PostQuitMessage(0);
                LRESULT(0)
            }
            _ => DefWindowProcW(hwnd, message, wparam, lparam),
        }
    }
}

// Function to show or hide the window
pub unsafe fn set_window_visibility(hwnd: HWND, visible: bool) {
    if hwnd.0 != 0 {
        if visible {
            ShowWindow(hwnd, SW_SHOW);
            info!("Window shown (streaming active)");
        } else {
            ShowWindow(hwnd, SW_HIDE);
            info!("Window hidden (not streaming)");
        }
    }
}